### DISCUSSION TITLE
Starting vim/neovim **after** zsh initialisation

### DISCUSSION THREAD

**Discussion by anjos** [kovidgoyal/kitty#6019] [Q&A] [answered]
I have some initialisation on my zsh (c.f. direnv support) that allows me to automatically switch working (virtual) environments as I enter a new directory.  When I start Neovim from a shell in a certain working directory, it is setup to take the currently set virtual environment as the default environment for auto-completions and type checking.  This works reliably.

I wanted to use kitty's launch command to launch a new window/tab, at the current environment, with a vim editor running.  So I tried:

```
map cmd+y launch --cwd=current --type=window nvim
```

That does launch Neovim, however my (shell) initialisation is missing, therefore the direnv support.

I suspect this must be trivial, but I can't seem to get a working command...  This does not work at all:

```
map cmd+y launch --cwd=current --type=window zsh -i -c exec nvim
```

Does anybody here know a solution for this?  Thanks a lot in advance!

**Comment by kovidgoyal** [comment]
See man zsh when you run zsh with -c, zsh does not source ~/.zshrc. You will
need to source it manually.

**Comment by anjos** [comment]
I will try that - thanks a lot!  I also saw a feature called `clone-in-kitty` or `edit-in-kitty`, which may allow me to clone the current environment of the running shell, and launch an editor in a separate tab or window.  Can that be a more efficient replacement just to launch a new vi session with the current setup?  If so, how to hook it to key keybinding such as the above?

**Reply by anjos** [reply]
I mean, the manual mentions it as some sort of drop-in replacement for launch, with some limitations, but this does not seem to work, nor have I found any documentation explaining how to bind that action to a key:

```
map cmd+y clone-in-kitty --type=window --location=before nvim
```

Too naive?

**Reply by page-down** [reply]
In order to clone the environment, e.g. restore almost all environment variables in the shell of a new window, activate the respective virtual environment, access to the shell's own environment is required, so this is a function in the shell that needs to be typed at the shell prompt.

And this may conflict with your own automatic venv activation, as this will go to the current working directory and load the shell in full.

> I suspect this must be trivial, ...

Bind the shortcut keys to execute a shell script, prepare the environment variables you need and activate venv then execute vim.
Based on what you have described, it is that straightforward.

Of course there is another way, which is to execute vim in your shell rc or environment activated shell script.

**Reply by anjos** [reply]
OK, thanks for the clarification this is a shell-integration feature.  All makes sense.

**Comment by anjos** [comment]
I thought I'd post a working solution here, following your advice.  I settled on creating a zsh script that does the necessary to setup.  I put this script in `${XDG_CONFIG_HOME}/kitty/nvim.zsh`:

```sh
# Adds homebrew setup
eval "$(/opt/homebrew/bin/brew shellenv)"

# Integrates direnv
eval "$(direnv hook zsh)"

# Triggers reloading of direnv's .envrc
cd ~ && cd -

exec nvim "$@"
```

... and invoke it with `cmd+d` like so:

```
map cmd+d launch --cwd=current --type=window zsh ~/.config/kitty/nvim.zsh
```

This works as expected, to launch a window where Neovim will run, with a configured environment.

**Reply by anjos** [reply]
Of course, if you have any simplification/improvement suggestions, I'm all ears.

**Reply by page-down** [reply]
```zsh
# Triggers reloading of direnv's .envrc
cd ~ && cd -
```

I think `direnv exec` can do something similar without switching working directory twice.
If you don't use both Intel and Apple Silicon Macs, I think the contents of `brew shellenv` will only be rarely changed.

**Reply by anjos** [reply]
Beautiful.  That makes it simpler:

```sh
# Adds homebrew setup
eval "$(/opt/homebrew/bin/brew shellenv)"

# Integrates direnv
eval "$(direnv hook zsh)"

# Runs editor inside preset environment
direnv exec . nvim "$@"
```

If the path to `direnv` is known one could also set `exe_search_path`:

```
exe_search_path /opt/homebrew/bin
```

... which would reduce the launch script to a one-liner:

```sh
eval "$(direnv hook zsh)" && direnv exec . nvim "$@"
```

I haven't managed to integrate that one line into kitty's `kitty.conf` though:

```
exe_search_path /opt/homebrew/bin
map cmd+d launch --cwd=current --type=window zsh 'eval "$(direnv hook zsh)" && direnv exec . nvim'
```

... doesn't really work.  Clues?

**Reply by page-down** [reply]
The option `exe_search_path` will only be used to find commands launched directly by kitty, such as `nvim` after `launch`.

You need `env` to specify the `PATH` that kitty provides to the child process being started. Note that this `PATH` will not be used for `launch` to find commands.

```conf
env PATH=$HOME/.local/bin:/opt/homebrew/bin:$PATH
exe_search_path /opt/homebrew/bin
```

> map launch ... zsh 'eval ...'

You need `zsh -c` to execute the argument as code, otherwise this will not work.

UPDATE:
The last command can be `exec direnv exec ..... `, so that zsh will not run in the background all the time.
You can also execute `...; direnv exec ... ; zsh -il`, which will return to zsh after exiting nvim instead of closing the window, if you want.

**Comment by anjos** [comment, marked as answer]
Thanks for all these tips.  The final solution fits inside `kitty.conf`, and looks like this:

```
# launches nvim after sourcing current .envrc (direnv)
env PATH=/opt/homebrew/bin:$PATH
exe_search_path /opt/homebrew/bin
map cmd+e launch --cwd=current --type=window zsh -c 'eval "$(direnv hook zsh)"; exec direnv exec . nvim'
```
